File: Solution/0006.cpp
// Complexity: O(N)

class Solution {
public:
    string convert(string s, int numRows) {
        if (numRows == 1)
            return s;
            
        vector<string> out;
        for (int i=0; i<numRows; ++i)
            out.push_back("");
        
        int returnNum = 2 * numRows - 2;
        for (int i=0; i<s.length(); ++i) {
            if (i % returnNum < numRows)
                out[i % returnNum] += s[i];
            else {
                out[returnNum - (i % returnNum)] += s[i];
            }
        }
        
        string out_string = "";
        for (int i=0; i<out.size(); ++i) 
            out_string += out[i];
        
        return out_string;
    }
};

File: Solution/0006.py
class Solution:
    def convert(self, s: str, numRows: int) -> str:
        n = len(s)
        tmp = ["" for _ in range(numRows)]
        if numRows == 1:
            return s

        repeat_num = (numRows - 1) * 2
        for i in range(len(s)):
            if i % repeat_num < numRows:
                tmp[i % repeat_num] += s[i]
            else:
                tmp[repeat_num - (i % repeat_num)] += s[i]
        
        result = ""
        for tmp_str in tmp:
            print(tmp_str)
            result += tmp_str
        
        return result